### PR TITLE
Fix tests without assertions

### PR DESCRIPTION
**Pull request by gsteel**
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | no
| RFC           | no
| QA            | yes

### Description

A number of the tests _(Around 30)_ complain about no assertions which is pretty noisy when you're running tests locally.

This pull updates those tests to use `$this->expectedException()` where appropriate instead of try/fail and add `$this->addToAssertionCount()` when appropriate.

There's also a little bit of de-duplication surrounding the escape helpers and permitted encodings and updates to mock usage so that the mocks assert some expectations where previously there were none.

### CODE REVIEW

File: test/Helper/EscaperEncodingsTrait.php [hunk 1]
```diff
@@ -0,0 +1,29 @@
+<?php
+
+declare(strict_types=1);
```
Comment (froschdesign) [inline test/Helper/EscaperEncodingsTrait.php]: Please add the licence header for this file.

File: test/Helper/HeadLinkTest.php [hunk 2]
```diff
@@ -319,18 +321,24 @@ public function testConditionalStylesheetCreationNoIEWidthSpaces()
         $this->assertStringContainsString('<!--<![endif]-->', $string);
     }
 
-    public function testSettingAlternateWithTooFewArgsRaisesException()
+    public function argumentCountProvider() : iterable
```
Comment (froschdesign) [inline test/Helper/HeadLinkTest.php]: Please run the coding standard check. There should be no space before the colon.

File: test/Helper/HeadLinkTest.php [hunk 3]
```diff
@@ -319,18 +321,24 @@ public function testConditionalStylesheetCreationNoIEWidthSpaces()
...
-            $this->fail('Setting alternate with fewer than 3 args should raise exception');
-        } catch (ViewException $e) {
-        }
+        return [
+            'One' => [1],
+            'Two' => [2],
+        ];
+    }
+
+    /** @dataProvider argumentCountProvider */
```
Comment (froschdesign) [inline test/Helper/HeadLinkTest.php:332]: Please run the coding standard check. I don't know if the short DocBlock is allowed for methods.

**Comment by gsteel** [comment]
Whoops… Just rebased on 2.12.x - was I wrong to do that?

File: test/Helper/HeadMetaTest.php [hunk 4]
```diff
@@ -77,27 +77,34 @@ public function testHeadMetaReturnsObjectInstance()
...
     }
 
-    public function testAppendPrependAndSetThrowExceptionsWhenNonMetaValueProvided()
+    public function testAppendPrependAndSetThrowExceptionsWhenNonMetaValueProvided(): void
     {
         try {
             $this->helper->append('foo');
             $this->fail('Non-meta value should not append');
         } catch (ViewException $e) {
+            $this->addToAssertionCount(1);
```
Comment (froschdesign) [inline test/Helper/HeadMetaTest.php]: Please test here that an exception is thrown.

File: test/Helper/HeadScriptTest.php [hunk 5]
```diff
@@ -72,21 +72,28 @@ public function testSetPrependAppendAndOffsetSetThrowExceptionsOnInvalidItems()
             $this->helper->append('foo');
             $this->fail('Append should throw exception with invalid item');
         } catch (View\Exception\ExceptionInterface $e) {
+            $this->addToAssertionCount(1);
```
Comment (froschdesign) [inline test/Helper/HeadScriptTest.php]: The same here, test that an exception is thrown.

File: test/Helper/Navigation/BreadcrumbsTest.php [hunk 6]
```diff
@@ -233,17 +233,14 @@ public function testRenderingPartialBySpecifyingAnArrayAsPartial()
...
-            $this->fail(
-                '$partial was invalid, but no Laminas\View\Exception\ExceptionInterface was thrown'
-            );
-        } catch (ExceptionInterface $e) {
-        }
+        $this->expectException(InvalidArgumentException::class);
+        $this->_helper->render();
+        $this->fail(
+            '$partial was invalid, but no Laminas\View\Exception\ExceptionInterface was thrown'
+        );
```
Comment (froschdesign) [inline test/Helper/Navigation/BreadcrumbsTest.php]: I think calling the fail is not needed anymore.

File: test/Helper/PartialLoopTest.php [hunk 7]
```diff
@@ -227,13 +226,14 @@ public function testShouldNotCastToArrayIfObjectIsTraversable()
...
      */
-    public function testEmptyArrayPassedToPartialLoopShouldNotThrowException()
+    public function testEmptyArrayPassedToPartialLoopShouldNotThrowException(): void
     {
         $view = new View();
         $view->resolver()->addPath($this->basePath . '/application/views/scripts');
         $this->helper->setView($view);
 
         $this->helper->__invoke('partialLoop.phtml', []);
+        $this->addToAssertionCount(1);
```
Comment (froschdesign) [inline test/Helper/PartialLoopTest.php]: I think the partial counter can be compared here.

**Review by froschdesign — CHANGES_REQUESTED** [review]
Thanks for your contribution! 😃 

I added some suggestions for updates. Thanks in advance! 👍

**Comment by froschdesign** [comment]
@gsteel 
> Whoops… Just rebased on 2.12.x - was I wrong to do that?

This will be shipped as a bugfix so the branch for latest version is correct.

File: test/Helper/HeadLinkTest.php (hunk 2 above)
Comment (gsteel) [reply test/Helper/HeadLinkTest.php]: CS checks pass - you're right of course, but for some reason it's not picked up. Coding standard 1.x is in use - perhaps another pull to update it to ~2 at some point?

File: test/Helper/Navigation/BreadcrumbsTest.php (hunk 6 above)
Comment (gsteel) [reply test/Helper/Navigation/BreadcrumbsTest.php]: No… it's not needed but I thought leaving it would aid docs ?

Comment (froschdesign) [reply test/Helper/Navigation/BreadcrumbsTest.php]: The call of `expectException` is enough already.

File: test/Helper/HeadLinkTest.php (hunk 2 above)
Comment (froschdesign) [reply test/Helper/HeadLinkTest.php]: The new continuous integration workflow must be added before. A new pull request needed like this one: https://github.com/laminas/laminas-paginator/pull/18

File: test/Helper/HeadLinkTest.php (hunk 3 above)
Comment (gsteel) [reply test/Helper/HeadLinkTest.php:332]: Just checked this with coding-std ~2 and the short doc block is preferred. ~1 doesn't complain.

File: test/Helper/HeadScriptTest.php [hunk 8]
```diff
@@ -216,27 +220,27 @@ public function testOverloadOffsetSetScriptWritesToSpecifiedIndex()
...
     }
 
-    public function testOverloadingWithTooFewArgumentsRaisesException()
+    public function testOverloadingWithTooFewArgumentsRaisesException(): void
     {
         try {
             $this->helper->setScript();
             $this->fail('Too few arguments should raise exception');
         } catch (View\Exception\ExceptionInterface $e) {
+            $this->addToAssertionCount(1);
```
Comment (froschdesign) [inline test/Helper/HeadScriptTest.php]: Please remove all `addToAssertionCount` as used here.

File: test/Helper/Navigation/MenuTest.php [hunk 9]
```diff
@@ -294,15 +295,12 @@ public function testRenderingPartialWithParams()
...
         $this->_helper->setPartial(['menu.phtml']);
-
-        try {
-            $this->_helper->render();
-            $this->fail('invalid $partial should throw Laminas\View\Exception\InvalidArgumentException');
-        } catch (\Laminas\View\Exception\ExceptionInterface $e) {
-        }
+        $this->expectException(InvalidArgumentException::class);
+        $this->_helper->render();
+        $this->fail('invalid $partial should throw Laminas\View\Exception\InvalidArgumentException');
```
Comment (froschdesign) [inline test/Helper/Navigation/MenuTest.php]: Please remove also all `fail`.

**Review by Ocramius — APPROVED** [review]
LGTM! Will aggregate related patches together, so we can see if it all goes green after merge :)

**Pull request merged**